uefi/src/table/boot.rs

```diff
@@ -1604,14 +1623,173 @@ impl Align for MemoryDescriptor {
 #[repr(C)]
 pub struct MemoryMapKey(usize);
 
-/// A structure containing the size of a memory descriptor and the size of the
-/// memory map.
+/// The backing memory for the UEFI memory app on the UEFI heap, allocated using
+/// the UEFI boot services allocator. This occupied memory will also be
+/// reflected in the memory map itself.
+///
+/// Although untyped, it is similar to the `Box` type in terms of heap
+/// allocation and deallocation, as well as ownership of the corresponding
+/// memory. Apart from that, this type only has the semantics of a buffer.
+///
+/// The memory is untyped, which is necessary due to the nature of the UEFI
+/// spec. It still ensures a correct alignment to hold [`MemoryDescriptor`]. The
+/// size of the buffer is sufficient to hold the memory map at the point in time
+/// where this is created. Note that due to (not obvious or asynchronous)
+/// allocations/allocations in your environment, this might be outdated at the
+/// time you store the memory map in it.
+///
+/// Note that due to the nature of the UEFI memory app, this buffer might
+/// hold (a few) bytes more than necessary. The `map_size` reported by
+/// `get_memory_map` tells the actual size.
+///
+/// When this type is dropped and boot services are not exited yet, the memory
+/// is freed.
+///
+/// # Usage
+/// The type is intended to be used like this:
+/// 1. create it using [`MemoryMapBackingMemory::new`]
+/// 2. pass it to [`BootServices::get_memory_map`]
+/// 3. construct a [`MemoryMap`] from it
 #[derive(Debug)]
-pub struct MemoryMapSize {
-    /// Size of a single memory descriptor in bytes
-    pub entry_size: usize,
-    /// Size of the entire memory map in bytes
+#[allow(clippy::len_without_is_empty)] // this type is never empty
+pub(crate) struct MemoryMapBackingMemory(NonNull<[u8]>);
```

Comment:
To keep the code and API simpler, can we fold this into `MemoryMap` and drop the separate `MemoryMapBackingMemory` struct? `MemoryMap.buf` can be a `NonNull<[u8]>`, and `MemoryMap::new` would have more or less the same implementation that's currently in `MemoryMapBackingMemory::new`, and the `Drop` implementation would also move to `MemoryMap`.

Reply:
I think using a type state pattern might solve this:

`MemoryMap<'uninitialized>`

and `MemoryMap<'initialized>`

but then we export internal functionality to a user... I'd prefer if a user just sees `MemoryMap` which is already perfectly valid and initialized

Reply:
I see what you mean, having a separate type here may well be easiest. And anyway, it's internal to the crate so we can always tweak it later without affecting users.